File: libs/wasm-sdk/wasm-sdk-kotlin/src/wasmWasiMain/kotlin/sdk/Context.kt
package sdk

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.double
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.long
import kotlinx.serialization.json.longOrNull

class Context(private val input: ExecutionInput) {
    private val result = ExecutionResult.success()
    private val outputs = mutableMapOf<String, String>()

    // -- Metadata --

    val nodeId: String get() = input.nodeId
    val nodeName: String get() = input.nodeName
    val runId: String get() = input.runId
    val appId: String get() = input.appId
    val boardId: String get() = input.boardId
    val userId: String get() = input.userId
    val streamEnabled: Boolean get() = input.streamState
    val logLevelValue: Int get() = input.logLevel

    // -- Input getters --

    fun getString(name: String, default: String = ""): String {
        val element = input.inputs[name] ?: return default
        if (element is JsonPrimitive) {
            return element.contentOrNull ?: default
        }
        return default
    }

    fun getI64(name: String, default: Long = 0L): Long {
        val element = input.inputs[name] ?: return default
        if (element is JsonPrimitive) {
            return element.longOrNull ?: default
        }
        return default
    }

    fun getF64(name: String, default: Double = 0.0): Double {
        val element = input.inputs[name] ?: return default
        if (element is JsonPrimitive) {
            return element.doubleOrNull ?: default
        }
        return default
    }

    fun getBool(name: String, default: Boolean = false): Boolean {
        val element = input.inputs[name] ?: return default
        if (element is JsonPrimitive) {
            return element.booleanOrNull ?: default
        }
        return default
    }

    fun getRawInput(name: String): String? {
        val element = input.inputs[name] ?: return null
        return Json.encodeToString(kotlinx.serialization.json.JsonElement.serializer(), element)
    }

    // -- Output setters --

    fun setOutput(name: String, value: String) {
        outputs[name] = Json.encodeToString(kotlinx.serialization.json.JsonElement.serializer(), JsonPrimitive(value))
    }

    fun setOutput(name: String, value: Long) {
        outputs[name] = value.toString()
    }

    fun setOutput(name: String, value: Double) {
        outputs[name] = value.toString()
    }

    fun setOutput(name: String, value: Boolean) {
        outputs[name] = value.toString()
    }

    fun setOutputJson(name: String, json: String) {
        outputs[name] = json
    }

    fun activateExec(pinName: String) {
        result.activateExec.add(pinName)
    }

    fun setPending(pending: Boolean) {
        result.pending = pending
    }

    fun setError(error: String) {
        result.error = error
    }

    // -- Level-gated logging --

    private fun shouldLog(level: Int): Boolean = level >= input.logLevel

    fun debug(message: String) {
        if (shouldLog(LogLevel.DEBUG)) logDebug(message)
    }

    fun info(message: String) {
        if (shouldLog(LogLevel.INFO)) logInfo(message)
    }

    fun warn(message: String) {
        if (shouldLog(LogLevel.WARN)) logWarn(message)
    }

    fun error(message: String) {
        if (shouldLog(LogLevel.ERROR)) logError(message)
    }

    // -- Conditional streaming --

    fun streamText(text: String) {
        if (streamEnabled) sdk.streamText(text)
    }

    fun streamJson(data: String) {
        if (streamEnabled) stream("json", data)
    }

    fun streamProgress(progress: Float, message: String) {
        if (streamEnabled) stream("progress", """{"progress":$progress,"message":"$message"}""")
    }

    // -- Variables --

    fun getVariable(name: String): String? = sdk.getVariable(name)

    fun setVariable(name: String, value: String) = sdk.setVariable(name, value)

    // -- Finalize --

    fun finish(): ExecutionResult {
        for ((name, value) in outputs) {
            val element = Json.decodeFromString(kotlinx.serialization.json.JsonElement.serializer(), value)
            result.outputs[name] = element
            hostSetOutput(name, value)
        }
        return result
    }

    fun success(): ExecutionResult {
        activateExec("exec_out")
        return finish()
    }

    fun fail(error: String): ExecutionResult {
        setError(error)
        return finish()
    }

    private fun hostSetOutput(name: String, value: String) {
        sdk.setOutput(name, value)
    }
}

File: libs/wasm-sdk/wasm-sdk-assemblyscript/assembly/context.ts
import { ExecutionInput, ExecutionResult, LogLevel } from "./types";
import {
  jsonString,
  parseJsonStringArray, parseJsonI64Array, parseJsonF64Array, parseJsonBoolArray, parseJsonStringMap,
  serializeStringArray, serializeI64Array, serializeF64Array, serializeBoolArray, serializeStringMap,
} from "./json";
import {
  debug as logDebug,
  info as logInfo,
  warn as logWarn,
  error as logError,
  stream,
  streamText as hostStreamText,
  getVariable as hostGetVariable,
  setVariable as hostSetVariable,
} from "./host";

export class Context {
  private _input: ExecutionInput;
  private _result: ExecutionResult;
  private _outputs: Map<string, string>;

  constructor(input: ExecutionInput) {
    this._input = input;
    this._result = ExecutionResult.success();
    this._outputs = new Map();
  }

  // -- Metadata --

  get nodeId(): string { return this._input.node_id; }
  get nodeName(): string { return this._input.node_name; }
  get runId(): string { return this._input.run_id; }
  get appId(): string { return this._input.app_id; }
  get boardId(): string { return this._input.board_id; }
  get userId(): string { return this._input.user_id; }
  get streamEnabled(): bool { return this._input.stream_state; }
  get logLevel(): u8 { return this._input.log_level; }

  // -- Input getters --

  getInput(name: string): string | null {
    return this._input.inputs.has(name) ? this._input.inputs.get(name) : null;
  }

  getString(name: string, defaultValue: string = ""): string {
    const val = this.getInput(name);
    if (val === null) return defaultValue;
    const v = val!;
    if (v.startsWith('"') && v.endsWith('"')) return v.slice(1, v.length - 1);
    return v;
  }

  getI64(name: string, defaultValue: i64 = 0): i64 {
    const val = this.getInput(name);
    if (val === null) return defaultValue;
    return I64.parseInt(val!);
  }

  getF64(name: string, defaultValue: f64 = 0): f64 {
    const val = this.getInput(name);
    if (val === null) return defaultValue;
    return F64.parseFloat(val!);
  }

  getBool(name: string, defaultValue: bool = false): bool {
    const val = this.getInput(name);
    if (val === null) return defaultValue;
    return val! == "true";
  }

  requireInput(name: string): string {
    const val = this.getInput(name);
    if (val === null) throw new Error("Missing required input: " + name);
    return val!;
  }

  // -- Array input getters --

  getStringArray(name: string): string[] {
    const val = this.getInput(name);
    if (val === null) return [];
    return parseJsonStringArray(val!);
  }

  getI64Array(name: string): i64[] {
    const val = this.getInput(name);
    if (val === null) return [];
    return parseJsonI64Array(val!);
  }

  getF64Array(name: string): f64[] {
    const val = this.getInput(name);
    if (val === null) return [];
    return parseJsonF64Array(val!);
  }

  getBoolArray(name: string): bool[] {
    const val = this.getInput(name);
    if (val === null) return [];
    return parseJsonBoolArray(val!);
  }

  // -- Map input getters --

  getStringMap(name: string): Map<string, string> {
    const val = this.getInput(name);
    if (val === null) return new Map<string, string>();
    return parseJsonStringMap(val!);
  }

  // -- Output setters --

  setOutput(name: string, value: string): void {
    this._outputs.set(name, value);
  }

  setString(name: string, value: string): void {
    this._outputs.set(name, jsonString(value));
  }

  setI64(name: string, value: i64): void {
    this._outputs.set(name, value.toString());
  }

  setF64(name: string, value: f64): void {
    this._outputs.set(name, value.toString());
  }

  setBool(name: string, value: bool): void {
    this._outputs.set(name, value ? "true" : "false");
  }

  // -- Array output setters --

  setStringArray(name: string, values: string[]): void {
    this._outputs.set(name, serializeStringArray(values));
  }

  setI64Array(name: string, values: i64[]): void {
    this._outputs.set(name, serializeI64Array(values));
  }

  setF64Array(name: string, values: f64[]): void {
    this._outputs.set(name, serializeF64Array(values));
  }

  setBoolArray(name: string, values: bool[]): void {
    this._outputs.set(name, serializeBoolArray(values));
  }

  // -- Map output setters --

  setStringMap(name: string, values: Map<string, string>): void {
    this._outputs.set(name, serializeStringMap(values));
  }

  activateExec(pinName: string): void {
    this._result.activate_exec.push(pinName);
  }

  setPending(pending: bool): void {
    this._result.pending = pending;
  }

  setError(error: string): void {
    this._result.error = error;
  }

  // -- Level-gated logging --

  shouldLog(level: LogLevel): bool {
    return level >= this._input.log_level;
  }

  debug(message: string): void {
    if (this.shouldLog(LogLevel.Debug)) logDebug(message);
  }

  info(message: string): void {
    if (this.shouldLog(LogLevel.Info)) logInfo(message);
  }

  warn(message: string): void {
    if (this.shouldLog(LogLevel.Warn)) logWarn(message);
  }

  error(message: string): void {
    if (this.shouldLog(LogLevel.Error)) logError(message);
  }

  // -- Conditional streaming --

  streamText(text: string): void {
    if (this.streamEnabled) hostStreamText(text);
  }

  streamJson(data: string): void {
    if (this.streamEnabled) stream("json", data);
  }

  streamProgress(progress: f32, message: string): void {
    if (this.streamEnabled) stream("progress", `{"progress":${progress},"message":"${message}"}`);
  }

  // -- Variables --

  getVariable(name: string): string | null {
    return hostGetVariable(name);
  }

  setVariable(name: string, value: string): void {
    hostSetVariable(name, value);
  }

  // -- Finalize --

  finish(): ExecutionResult {
    const keys = this._outputs.keys();
    for (let i = 0; i < keys.length; i++) {
      this._result.setOutput(keys[i], this._outputs.get(keys[i]));
    }
    return this._result;
  }

  success(): ExecutionResult {
    this.activateExec("exec_out");
    return this.finish();
  }

  fail(error: string): ExecutionResult {
    this.setError(error);
    return this.finish();
  }
}
